### PR TITLE
Introduce Coupa and Pardot connector support

### LINKED ISSUE #459: Add Coupa and Pardot connector support
We need to add support for Coupa and Pardot connector.

### PR DESCRIPTION
**Pull request by peter-plochan**
Fixes #459

### CODE REVIEW

File: src/main/java/com/gooddata/connector/CoupaInstances.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+package com.gooddata.connector;
+
+import com.fasterxml.jackson.annotation.*;
+
+import java.util.*;
+
+/**
+ * Internal only. Deserialization only.
+ * Contains map (apiUrl -> CoupaInstance) of coupa instances.
```
Comment (liry) [inline src/main/java/com/gooddata/connector/CoupaInstances.java]: really map?

**Comment by liry** [comment]
the rest LGTM

File: src/main/java/com/gooddata/connector/CoupaInstances.java (hunk 1 above)
Comment (peter-plochan) [inline src/main/java/com/gooddata/connector/CoupaInstances.java]: fixed

File: src/main/java/com/gooddata/connector/ConnectorService.java [hunk 2]
```diff
@@ -147,6 +147,55 @@ public Zendesk4Settings getZendesk4Settings(Project project) {
...
+            throw new ConnectorException("Unable to create Coupa instance with API URL '" + instance.getApiUrl() +
+                    "'", e);
+        }
+    }
+
+    /**
+     * Returns collection of all existing Coupa instances
+     *
+     * @param project project
+     * @return collection of Coupa instances
```
Comment (martiner) [inline src/main/java/com/gooddata/connector/ConnectorService.java]: document and ensure the return value is never null

File: src/main/java/com/gooddata/connector/ConnectorService.java [hunk 3]
```diff
@@ -147,6 +147,55 @@ public Zendesk4Settings getZendesk4Settings(Project project) {
...
+     */
+    public Collection<CoupaInstance> getCoupaInstances(Project project) {
```
Comment (martiner) [inline src/main/java/com/gooddata/connector/ConnectorService.java]: naming: `findCoupaInstances()` https://github.com/gooddata/gooddata-java/blob/master/CONTRIBUTING.md#services

File: src/main/java/com/gooddata/connector/CoupaInstance.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+     * @param name instance name
+     * @param apiUrl API URL
+     * @param apiKey API key for this instance
+     */
+    @JsonCreator
+    public CoupaInstance(
+            @JsonProperty("name") String name,
+            @JsonProperty("apiUrl") String apiUrl,
+            @JsonProperty("apiKey") String apiKey) {
+        this.name = name;
```
Comment (martiner) [inline src/main/java/com/gooddata/connector/CoupaInstance.java]: validations?

File: src/main/java/com/gooddata/connector/CoupaProcessExecution.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    private Boolean incremental;
+    private LocalDate downloadDataFrom;
+
+    @Override
+    public ConnectorType getConnectorType() {
+        return ConnectorType.COUPA;
+    }
+
+    @JsonProperty("incremental")
```
Comment (martiner) [inline src/main/java/com/gooddata/connector/CoupaProcessExecution.java]: seems redundant

File: src/main/java/com/gooddata/connector/CoupaProcessExecution.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        return incremental;
+    }
+
+    @JsonIgnore
+    public void setIncremental(Boolean incremental) {
+        this.incremental = incremental;
+    }
+
+    @JsonProperty("downloadDataFrom")
+    @JsonSerialize(using = GDDateSerializer.class)
```
Comment (martiner) [inline src/main/java/com/gooddata/connector/CoupaProcessExecution.java]: use serializers in constructor

File: src/main/java/com/gooddata/connector/CoupaProcessExecution.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    public ConnectorType getConnectorType() {
+        return ConnectorType.COUPA;
+    }
+
+    @JsonProperty("incremental")
+    public Boolean getIncremental() {
+        return incremental;
+    }
+
+    @JsonIgnore
```
Comment (martiner) [inline src/main/java/com/gooddata/connector/CoupaProcessExecution.java]: why? javadoc states it is serialization only

**Pull request merged**